cirq-core/cirq/circuits/circuit_operation.py

```diff
@@ -226,38 +226,19 @@ def mapped_circuit(self, deep: bool = False) -> 'cirq.Circuit':
         if has_measurements:
             circuit = protocols.with_measurement_key_mapping(circuit, self.measurement_key_map)
         circuit = protocols.resolve_parameters(circuit, self.param_resolver, recursive=False)
-        if deep:
-
-            def map_deep(op: 'cirq.Operation') -> 'cirq.OP_TREE':
-                return op.mapped_circuit(deep=True) if isinstance(op, CircuitOperation) else op
-
-            if self.repetition_ids is None:
-                return circuit.map_operations(map_deep)
+        if self.repetition_ids:
             if not has_measurements:
-                return circuit.map_operations(map_deep) * abs(self.repetitions)
-
-            # Path must be constructed from the top down.
-            rekeyed_circuit = circuits.Circuit(
-                protocols.with_key_path(circuit, self.parent_path + (rep,))
-                for rep in self.repetition_ids
+                circuit = circuit * abs(self.repetitions)
+            else:
+                circuit = circuits.Circuit(
+                    protocols.with_key_path(circuit, self.parent_path + (rep,))
+                    for rep in self.repetition_ids
+                )
+        if deep:
+            circuit = circuit.map_operations(
+                lambda op: op.mapped_circuit(deep=True) if isinstance(op, CircuitOperation) else op
```

Comment:
There's a reason behind the previous "mapped_circuit then repetitions" ordering: if we unroll the repetitions first, we have to remap each repetition separately. This can become costly for dense subcircuit nesting schemes.

Reply:
Fixed.

Reply:
Actually let's just abandon this and move straight to #4619. In that PR I can just swap the two `if` branches completely rather than splitting up the `if self.repetition_ids` into a before-and-after chunk like I'm doing here. I can't move the whole of that branch above the `map_deep` branch here because it causes the bug fixed in #4619 to occur.